File: RGBImage.java

package ce326.hw2;

import ce326.hw2.RGBPixel;
import java.awt.Graphics2D;
import java.awt.Image;
import java.io.File;
import java.io.IOException;
import java.lang.*;

import javax.imageio.ImageIO;

public class RGBImage implements ce326.hw2.Image, Cloneable{
    public int colordepth;
    public int width;
    public int height;
    RGBPixel[][] array;
    public final int MAX_COLORDEPTH = 255;
        
    public RGBImage(){//default constructor
    }
    
    public RGBImage(int width, int height, int colordepth){
        this.width = width;
        this.height = height;
        this.colordepth = colordepth;
        array = new RGBPixel[height][width];
        for(int i = 0; i < height; i++){
            for(int j = 0; j < width; j++){
                array[i][j] = new RGBPixel();
            }
        }
    }
    
    public RGBImage(RGBImage copyImg){//image copy of copyImg
        colordepth = copyImg.colordepth;
        height = copyImg.height;
        width = copyImg.width;
        array = new RGBPixel[height][width];
        array = copyImg.array;
    }
    
    public RGBImage(YUVImage YUVImg){
        colordepth = 255;
        height = YUVImg.height;
        width = YUVImg.width;
        
        array = new RGBPixel[height][width];
        for(int i = 0; i < height; i++){
            for(int j = 0; j < width; j++){
                array[i][j] = new RGBPixel(YUVImg.array[i][j]);
            }
        }
    
    }
    
    int getWidth(){
        return(width);
    }
    
    int getHeight(){
        return(height);
    }
    
    int getColorDepth(){
        return(colordepth);
    }
    
    RGBPixel getPixel(int row, int col){
        return array[row][col];
    }
    
    void setPixel(int row, int col,  RGBPixel pixel){        
        array[row][col]= pixel;
        array[row][col].Red = pixel.Red;
        array[row][col].Green = pixel.Green;
        array[row][col].Blue = pixel.Blue;
    }
    
    public void printRGB(){
         for(int i = 0; i < height; i++){
            for(int j = 0; j < width; j++){
                String fs = String.format("%s %s %s\n", array[i][j].Red, array[i][j].Green, array[i][j].Blue);
                System.out.println(fs);
            }
         }
    }
    
    public RGBImage clone() throws CloneNotSupportedException {
        RGBImage b = (RGBImage)super.clone();
        return b;
    }
    
    public void grayscale(){//in every pixel -> red*0.3, green*0.59, blue*0.11
        short gray = 0;//all 3 colours have the same value
        
        for(int i = 0; i < height; i++){
            for(int j = 0; j < width; j++){
                gray = (short) ((0.11 * array[i][j].Blue) + (0.3 * array[i][j].Red) + (0.59 * array[i][j].Green));
                array[i][j].Blue = gray;
                array[i][j].Red = gray;
                array[i][j].Green = gray;
            }
        }        
    }
   
    public void doublesize(){//brightness of each pixel is copied in different positions
        RGBPixel [][] arrayhelp = new RGBPixel[2*height][2*width];
        for(int i = 0; i < (height*2); i++){
            for(int j = 0; j < (width*2); j++){
               arrayhelp[i][j] = new RGBPixel((short)1, (short)2, (short)3);
            }
        }
        
        for(int i = 0; i < height; i++){
            for(int j = 0; j < width; j++){
                arrayhelp[2*i][2*j] = array[i][j];
                arrayhelp[2*i + 1][2*j] = array[i][j];
                arrayhelp[2*i][2*j + 1] = array[i][j];
                arrayhelp[2*i + 1][2*j + 1] = array[i][j];
            }   
        }
        array = arrayhelp.clone();
        height = 2*height;
        width = 2*width;
   }
   
    public void halfsize(){//brightness of a pixel = average brightness of other pixels
        int mesosOrosBlue = 0;
        int mesosOrosRed = 0;
        int mesosOrosGreen = 0;
        
        RGBPixel [][] arrayhelp = new RGBPixel[height/2][width/2];
        
        for(int i = 0; i < (height/2); i++){
            for(int j = 0; j < (width/2); j++){
               arrayhelp[i][j] = new RGBPixel((short)1, (short)2, (short)3);
            }
        }
               
        for(int i = 0; i < (height/2); i++){
            for(int j = 0; j < (width/2); j++){
                
                mesosOrosBlue += array[2*i][2*j].Blue;
                mesosOrosBlue += array[2*i + 1][2*j].Blue;
                mesosOrosBlue += array[2*i][2*j + 1].Blue;
                mesosOrosBlue += array[2*i + 1][2*j + 1].Blue;
                
                arrayhelp[i][j].Blue = (short) (mesosOrosBlue/4);
                
                mesosOrosRed += array[2*i][2*j].Red;
                mesosOrosRed += array[2*i + 1][2*j].Red;
                mesosOrosRed += array[2*i][2*j + 1].Red;
                mesosOrosRed += array[2*i + 1][2*j + 1].Red;
                
                arrayhelp[i][j].Red = (short) (mesosOrosRed/4);
                
                mesosOrosGreen += array[2*i][2*j].Green;
                mesosOrosGreen += array[2*i + 1][2*j].Green;
                mesosOrosGreen += array[2*i][2*j + 1].Green;
                mesosOrosGreen += array[2*i + 1][2*j + 1].Green;
                
                arrayhelp[i][j].Green = (short) (mesosOrosGreen/4);
                
                mesosOrosBlue = 0;
                mesosOrosRed = 0;
                mesosOrosGreen = 0;
            }
        }
        
        array = arrayhelp.clone();
        height = height/2;
        width = width/2;
        
   }
   
    public void rotateClockwise(){//rotate a photo 90 degrees clockwise
        RGBPixel[][] arrayHelp = new RGBPixel[width][height];
        int newHeight;
        int newColumn = 0;
        int newRow = 0;
        
        for(int i = 0; i < height; i++){
            for(int j = 0; j < width; j++){
                arrayHelp[j][height - i -1] = array[i][j];
            }
        }

        array = arrayHelp.clone();
        newHeight = width;
        width = height;
        height = newHeight;
        
    }

   
}

File: ce326/hw2/python/RGBImage.py
from RGBPixel import RGBPixel
import numpy as np

MAX_COLORDEPTH = 255

class RGBImage:

    def __init__(self, width, height, colordepth):#default constructor
        self.width = width;
        self.height = height;
        self.colordepth = colordepth;
        self.array = np.empty(shape = (self.width, self.height), dtype = RGBPixel);
        
        for i in range(self.height):
            for j in range (self.width):
                self.array[i][j] = RGBPixel(0, 0, 0, 0);
    

    def RGBImage_copy(self, copyImg): #image copy of copyImg
        self.colordepth = copyImg.colordepth;
        self.height = copyImg.height;
        self.width = copyImg.width;
        self.array.resize((copyImg.width, copyImg.height), refcheck = False);
        self.array = copyImg.array;

    def RGBImage_yuv(self, YUVImg):
        self.colordepth = 255;
        self.height = YUVImg.height;
        self.width = YUVImg.width;
     
        self.array = RGBPixel[self.height][self.width]();
        for i in range(self.height):
            for j in range (self.width):
                self.array[i][j] = RGBPixel(YUVImg.array[i][j]);

    def getWidth(self):
        return (self.width);

    def getHeight(self):
        return (self.height);

    def getColorDepth(self):
        return (self.colordepth);

    def getPixel(self,row, col):
        return self.array[row][col];

    def setPixel(self, row, col, pixel):
        self.array[row][col] = pixel;
        self.array[row][col].Red = pixel.Red;
        self.array[row][col].Green = pixel.Green;
        self.array[row][col].Blue = pixel.Blue;

    def printRGB(self):
        for i in range(self.height):
            for j in range (self.width):
                print(str(self.array[i][j].Red) + ", " + str(self.array[i][j].Green) + ", " + str(self.array[i][j].Blue));
                

    def grayscale(self): # in every pixel -> red*0.3, green*0.59, blue*0.11
        gray = 0; #all 3 colours have the same value
        for i in range(self.height):
            for j in range (self.width):
                 gray = (0.11 * self.array[i][j].Blue) + (0.3 * self.array[i][j].Red) + (0.59 * self.array[i][j].Green);
                 self.array[i][j].Blue = gray;
                 self.array[i][j].Red = gray;
                 self.array[i][j].Green = gray;
 

    def doublesize(self): #brightness of each pixel is copied in different positions
        arrayhelp = np.empty(shape = (2 * self.height, 2 * self.width), dtype = RGBPixel)
       
        for i in range(self.height * 2):
            for j in range (self.width * 2):
                arrayhelp[i][j] = RGBPixel(1, 2, 3, 15);
 
        for i in range(self.height):
            for j in range (self.width):
                 arrayhelp[2 * i][2 * j] = self.array[i][j];
                 arrayhelp[2 * i + 1][2 * j] = self.array[i][j];
                 arrayhelp[2 * i][2 * j + 1] = self.array[i][j];
                 arrayhelp[2 * i + 1][2 * j + 1] = self.array[i][j];
        
        
        self.array = arrayhelp;
        self.height = 2 * self.height;
        self.width = 2 * self.width;
        

    def foursize(self): #brightness of each pixel is copied in different positions
        arrayhelp = np.empty(shape = (4 * self.width, 4 * self.height), dtype = RGBPixel)
 
        for i in range(self.height * 4):
            for j in range (self.width * 4):
                arrayhelp[i][j] = RGBPixel(1, 2, 3, 15);

        for i in range(self.height):
            for j in range (self.width):
               arrayhelp[4 * i][4 * j] = self.array[i][j];
               arrayhelp[4 * i + 2][4 * j] = self.array[i][j];
               arrayhelp[4 * i][4 * j + 2] = self.array[i][j];
               arrayhelp[4 * i + 2][4 * j + 2] = self.array[i][j];
      
        self.array = arrayhelp;
        self.height = 4 * self.height;
        self.width = 4 * self.width;
    

    def halfsize(self): # brightness of a pixel = average brightness of other pixels
        mesosOrosBlue = 0;
        mesosOrosRed = 0;
        mesosOrosGreen = 0;

        arrayhelp = np.empty(shape = (int(self.height / 2), int(self.width / 2)), dtype = RGBPixel)

        for i in range(int(self.height / 2)):
            for j in range (int(self.width / 2)):
                arrayhelp[i][j] = RGBPixel(1, 2, 3, 15);

        for i in range(int(self.height / 2)):
            for j in range (int(self.width / 2)):
               mesosOrosBlue += self.array[2 * i][2 * j].Blue;
               mesosOrosBlue += self.array[2 * i + 1][2 * j].Blue;
               mesosOrosBlue += self.array[2 * i][2 * j + 1].Blue;
               mesosOrosBlue += self.array[2 * i + 1][2 * j + 1].Blue;
   
               arrayhelp[i][j].Blue = int(mesosOrosBlue / 4);
   
               mesosOrosRed += self.array[2 * i][2 * j].Red;
               mesosOrosRed += self.array[2 * i + 1][2 * j].Red;
               mesosOrosRed += self.array[2 * i][2 * j + 1].Red;
               mesosOrosRed += self.array[2 * i + 1][2 * j + 1].Red;
   
               arrayhelp[i][j].Red = int(mesosOrosRed / 4);
   
               mesosOrosGreen += self.array[2 * i][2 * j].Green;
               mesosOrosGreen += self.array[2 * i + 1][2 * j].Green;
               mesosOrosGreen += self.array[2 * i][2 * j + 1].Green;
               mesosOrosGreen += self.array[2 * i + 1][2 * j + 1].Green;
   
               arrayhelp[i][j].Green = int(mesosOrosGreen / 4);
   
               mesosOrosBlue = 0;
               mesosOrosRed = 0;
               mesosOrosGreen = 0;


        self.array = arrayhelp;
        self.height = int(self.height / 2);
        self.width = int(self.width / 2);


    def rotateClockwise(self): #rotate a photo 90 degrees clockwise
        arrayHelp = np.empty(shape = (self.width, self.height), dtype = RGBPixel)
        newColumn = 0;
        newRow = 0;

        for i in range(self.height):
            for j in range (self.width):
                arrayHelp[j][self.height - i - 1] = self.array[i][j];
 

        self.array = arrayHelp;
        newHeight = self.width;
        self.width = self.height;
        self.height = newHeight;
        
                 
                 
